File: services/java-microservices/product-service/src/main/java/com/ecommerce/productservice/controller/ProductController.java
package com.ecommerce.productservice.controller;

import com.ecommerce.productservice.dto.*;
import com.ecommerce.productservice.service.ProductService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.util.UUID;

@RestController
@RequestMapping("/api/v1/products")
@RequiredArgsConstructor
@Slf4j
@Tag(name = "Products", description = "Product management endpoints")
public class ProductController {

    private final ProductService productService;

    @PostMapping
    @Operation(summary = "Create a new product", description = "Creates a new product in the catalog")
    @ApiResponse(responseCode = "201", description = "Product created successfully")
    @ApiResponse(responseCode = "400", description = "Invalid input")
    @ApiResponse(responseCode = "409", description = "Product with same SKU or slug already exists")
    public ResponseEntity<ProductResponse> createProduct(@Valid @RequestBody ProductRequest request) {
        log.info("POST /api/v1/products - Creating product: {}", request.getName());
        ProductResponse response = productService.createProduct(request);
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @GetMapping("/{id}")
    @Operation(summary = "Get product by ID", description = "Retrieves a product by its unique identifier")
    @ApiResponse(responseCode = "200", description = "Product found")
    @ApiResponse(responseCode = "404", description = "Product not found")
    public ResponseEntity<ProductResponse> getProductById(
            @Parameter(description = "Product ID") @PathVariable UUID id) {
        log.info("GET /api/v1/products/{}", id);
        return ResponseEntity.ok(productService.getProductById(id));
    }

    @GetMapping
    @Operation(summary = "List products", description = "Lists all active products with pagination and filtering")
    @ApiResponse(responseCode = "200", description = "Products retrieved successfully")
    public ResponseEntity<PagedResponse<ProductResponse>> getAllProducts(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size,
            @RequestParam(defaultValue = "createdAt") String sortBy,
            @RequestParam(defaultValue = "desc") String sortDir,
            @RequestParam(required = false) UUID categoryId,
            @RequestParam(required = false) BigDecimal minPrice,
            @RequestParam(required = false) BigDecimal maxPrice,
            @RequestParam(required = false) Boolean featured,
            @RequestParam(required = false) String tag) {

        Sort sort = sortDir.equalsIgnoreCase("asc") ?
                Sort.by(sortBy).ascending() : Sort.by(sortBy).descending();
        Pageable pageable = PageRequest.of(page, size, sort);

        if (categoryId != null || minPrice != null || maxPrice != null || featured != null || tag != null) {
            return ResponseEntity.ok(productService.filterProducts(categoryId, minPrice, maxPrice, featured, tag, pageable));
        }

        return ResponseEntity.ok(productService.getAllProducts(pageable));
    }

    @PutMapping("/{id}")
    @Operation(summary = "Update product", description = "Updates an existing product")
    @ApiResponse(responseCode = "200", description = "Product updated successfully")
    @ApiResponse(responseCode = "404", description = "Product not found")
    public ResponseEntity<ProductResponse> updateProduct(
            @PathVariable UUID id,
            @Valid @RequestBody ProductRequest request) {
        log.info("PUT /api/v1/products/{}", id);
        return ResponseEntity.ok(productService.updateProduct(id, request));
    }

    @DeleteMapping("/{id}")
    @Operation(summary = "Delete product", description = "Soft deletes a product (sets is_active to false)")
    @ApiResponse(responseCode = "204", description = "Product deleted successfully")
    @ApiResponse(responseCode = "404", description = "Product not found")
    public ResponseEntity<Void> deleteProduct(@PathVariable UUID id) {
        log.info("DELETE /api/v1/products/{}", id);
        productService.deleteProduct(id);
        return ResponseEntity.noContent().build();
    }

    @GetMapping("/search")
    @Operation(summary = "Search products", description = "Full-text search across product name, description, and short description")
    @ApiResponse(responseCode = "200", description = "Search results retrieved")
    public ResponseEntity<PagedResponse<ProductResponse>> searchProducts(
            @Parameter(description = "Search query") @RequestParam String q,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size,
            @RequestParam(defaultValue = "name") String sortBy,
            @RequestParam(defaultValue = "asc") String sortDir) {

        Sort sort = sortDir.equalsIgnoreCase("asc") ?
                Sort.by(sortBy).ascending() : Sort.by(sortBy).descending();
        Pageable pageable = PageRequest.of(page, size, sort);

        log.info("GET /api/v1/products/search?q={}", q);
        return ResponseEntity.ok(productService.searchProducts(q, pageable));
    }

    @GetMapping("/category/{categoryId}")
    @Operation(summary = "Get products by category", description = "Retrieves all active products in a specific category")
    @ApiResponse(responseCode = "200", description = "Products retrieved successfully")
    @ApiResponse(responseCode = "404", description = "Category not found")
    public ResponseEntity<PagedResponse<ProductResponse>> getProductsByCategory(
            @PathVariable UUID categoryId,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size,
            @RequestParam(defaultValue = "createdAt") String sortBy,
            @RequestParam(defaultValue = "desc") String sortDir) {

        Sort sort = sortDir.equalsIgnoreCase("asc") ?
                Sort.by(sortBy).ascending() : Sort.by(sortBy).descending();
        Pageable pageable = PageRequest.of(page, size, sort);

        log.info("GET /api/v1/products/category/{}", categoryId);
        return ResponseEntity.ok(productService.getProductsByCategory(categoryId, pageable));
    }

    @GetMapping("/featured")
    @Operation(summary = "Get featured products", description = "Retrieves all featured products")
    @ApiResponse(responseCode = "200", description = "Featured products retrieved")
    public ResponseEntity<PagedResponse<ProductResponse>> getFeaturedProducts(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
        return ResponseEntity.ok(productService.getFeaturedProducts(pageable));
    }
}

File: services/nodejs-microservices/product-service/src/infrastructure/web/ProductController.ts
import { Router, Request, Response } from 'express';
import { ProductServiceImpl } from '../../application/ProductServiceImpl';
import { toProductResponse, toCategoryResponse, toReviewResponse } from './dto';

export class ProductController {
  public readonly router: Router;

  constructor(private readonly productService: ProductServiceImpl) {
    this.router = Router();
    this.setupRoutes();
  }

  private setupRoutes(): void {
    // Products
    this.router.post('/products', this.createProduct.bind(this));
    this.router.get('/products', this.listProducts.bind(this));
    this.router.get('/products/search', this.searchProducts.bind(this));
    this.router.get('/products/:id', this.getProductById.bind(this));
    this.router.put('/products/:id', this.updateProduct.bind(this));
    this.router.delete('/products/:id', this.deleteProduct.bind(this));

    // Categories
    this.router.post('/categories', this.createCategory.bind(this));
    this.router.get('/categories', this.listCategories.bind(this));

    // Reviews
    this.router.post('/products/:id/reviews', this.addReview.bind(this));
    this.router.get('/products/:id/reviews', this.getReviews.bind(this));
  }

  private async createProduct(req: Request, res: Response): Promise<void> {
    try {
      if (!req.body.name || !req.body.price) {
        res.status(400).json({ error: 'Missing required fields: name, price' });
        return;
      }
      const product = await this.productService.createProduct(req.body);
      res.status(201).json(toProductResponse(product));
    } catch (error: any) {
      console.error('[ProductController] Create error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  }

  private async getProductById(req: Request, res: Response): Promise<void> {
    try {
      const product = await this.productService.getProductById(req.params.id);
      if (!product) { res.status(404).json({ error: 'Product not found' }); return; }
      res.json(toProductResponse(product));
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  }

  private async listProducts(req: Request, res: Response): Promise<void> {
    try {
      const page = parseInt(req.query.page as string) || 1;
      const limit = parseInt(req.query.limit as string) || 20;
      const products = await this.productService.listProducts(page, limit);
      res.json(products.map(toProductResponse));
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  }

  private async searchProducts(req: Request, res: Response): Promise<void> {
    try {
      const query = req.query.q as string;
      if (!query) { res.status(400).json({ error: 'Query parameter "q" is required' }); return; }
      const products = await this.productService.searchProducts(query);
      res.json(products.map(toProductResponse));
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  }

  private async updateProduct(req: Request, res: Response): Promise<void> {
    try {
      const product = await this.productService.updateProduct(req.params.id, req.body);
      if (!product) { res.status(404).json({ error: 'Product not found' }); return; }
      res.json(toProductResponse(product));
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  }

  private async deleteProduct(req: Request, res: Response): Promise<void> {
    try {
      const deleted = await this.productService.deleteProduct(req.params.id);
      if (!deleted) { res.status(404).json({ error: 'Product not found' }); return; }
      res.status(204).send();
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  }

  private async createCategory(req: Request, res: Response): Promise<void> {
    try {
      if (!req.body.name) { res.status(400).json({ error: 'Name is required' }); return; }
      const category = await this.productService.createCategory(req.body);
      res.status(201).json(toCategoryResponse(category));
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  }

  private async listCategories(req: Request, res: Response): Promise<void> {
    try {
      const categories = await this.productService.listCategories();
      res.json(categories.map(toCategoryResponse));
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  }

  private async addReview(req: Request, res: Response): Promise<void> {
    try {
      if (!req.body.userId || !req.body.rating) {
        res.status(400).json({ error: 'Missing required fields: userId, rating' });
        return;
      }
      const review = await this.productService.addReview(req.params.id, req.body);
      res.status(201).json(toReviewResponse(review));
    } catch (error: any) {
      if (error.message === 'Product not found') {
        res.status(404).json({ error: error.message });
      } else {
        res.status(500).json({ error: 'Internal server error' });
      }
    }
  }

  private async getReviews(req: Request, res: Response): Promise<void> {
    try {
      const reviews = await this.productService.getReviews(req.params.id);
      res.json(reviews.map(toReviewResponse));
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  }
}
